refac: modularize kroma deploy

kangsorang opened the pull request:
Summary

This PR decouples the Kroma deployment logic from the Optimism deployment scripts, which were previously tightly coupled.

What’s changed
- Modularized Kroma deployment logic into a standalone library format

- Minimized code overlap with existing Optimism deploy code

- Improved maintainability and clarity by isolating Kroma-specific logic

Why it matters
By isolating Kroma deployment logic, we reduce interference with upstream Optimism code. This ensures smoother upstream merges and preserves the behavior of the original OP deployment flow.

kangsorang committed:
feat: implement Kroma deployment modules (ConfigBuilder, Deployer, Types, Initializers)
- Modularized Kroma contract deployment logic for better maintainability and reuse.

kangsorang committed:
feat: integrate Kroma deploy module into script & remove existing task code

kangsorang committed:
chore: adjust forge fmt

kangsorang committed:
feat: add kromaChainAssertions

Pangssu commented on packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:
Great. But would it be possible to separate this file as well? Just like DeployKroma.s.sol?

Pangssu commented on packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:
```diff
@@ -626,6 +621,18 @@ contract Deploy is Deployer {
+        vm.stopBroadcast();
+        vm.label(impl, string.concat(_name, "Impl"));
+    }
+
```
Was this not originally there?

kangsorang replied:
In DeployImplementation.s.sol, components are deployed using specific functions like deploySystemConfigImpl and deployL1CrossDomainMessengerImpl.
Here, deployment can be handled more generically, so a generalized function has been added.

Pangssu replied:
Since this is not part of the original Optimism code, it would be better not to define this function here.

kangsorang replied:
Ok, I’ll try to keep files like deploy.s.sol as close to the original Optimism version as possible, with minimal modifications.

Pangssu commented on packages/contracts-bedrock/scripts/deploy/kroma/KromaDeployer.sol:
```diff
@@ -0,0 +1,137 @@
+        // ========== Proxy Upgrade & Initialization ========== //
+        console.log("[Kroma Deploy] Upgrading and initializing proxies...");
+
+        vm.startBroadcast();
```
```suggestion
        vm.startBroadcast(msg.sender);
```

kangsorang replied:
Great, this allows us to separate the code much more cleanly.
However, it would be helpful to clarify what exactly defines a “Kroma contract.”
For example, even though OptimismPortal has been modified for Kroma, it’s not immediately clear whether it should be considered part of the Kroma-specific contract set.

kangsorang committed:
refactor: separate Kroma deploy logic from Optimism
- Refactored Kroma deployment into isolated script files, cleaned up logging style, and aligned structure with the existing Optimism deployment pattern.

Pangssu commented:
Rename the L2OutputOracle to KromaL2OutputOracle and rollback L2OO contract to the original version from optimism.

kangsorang commented:
> Rename the L2OutputOracle to KromaL2OutputOracle and rollback L2OO contract to the original version from optimism.

Great, this allows us to separate the code much more cleanly.
However, it would be helpful to clarify what exactly defines a “Kroma contract.”
For example, even though OptimismPortal has been modified for Kroma, it’s not immediately clear whether it should be considered part of the Kroma-specific contract set.

kangsorang commented:
> Rename the L2OutputOracle to KromaL2OutputOracle and rollback L2OO contract to the original version from optimism.

Additionally, if we rename the contract to KromaL2OutputOracle, the interface file should also follow the convention and become IKromaOutputOracle. However, many files currently reference IOutputOracle.sol or OutputOracle.sol, so this change would introduce a noticeable diff across the codebase. It’s definitely a trade-off.

Pangssu commented:
The ultimate goal is to avoid modifying Optimism's code.
If there are changes to the interface, the contract should be separated.
Minor modifications to the internal logic may be acceptable.

kangsorang commented:
> The ultimate goal is to avoid modifying Optimism's code. If there are changes to the interface, the contract should be separated. Minor modifications to the internal logic may be acceptable.

Sounds good — I’ll use interface changes and the degree of change (major vs. minor) as the main criteria.

kangsorang committed:
refactor: isolate KromaL2OutputOracle and preserve original L2OutputOracle

Pangssu commented on packages/contracts-bedrock/scripts/Artifacts.s.sol:
```diff
@@ -224,4 +224,11 @@ abstract contract Artifacts {
         initialized_ = uint8((uint256(slotVal) >> (slot.offset * 8)) & 0xFF);
     }
+
+    function setChainId(uint256 chainId) external {
+        console.log("set chainId to %s", chainId);
+        vm.chainId(chainId);
+        // update deploymentOutfile
+        deploymentOutfile = Config.deploymentOutfile();
+    }
```
Is it required?

kangsorang replied:
It’s needed in the setup contract for testing.
The code currently being used wasn’t included in this PR.
I’ll leave it out here and include it in a separate PR with a related topic.

dongchangYoo reviewed (approved):
LGTM
